sqlite3_opt_serialize.go

```diff
@@ -0,0 +1,88 @@
+// +build !libsqlite3 sqlite_serialize
+
+package sqlite3
+
+/*
+#ifndef USE_LIBSQLITE3
+#include <sqlite3-binding.h>
+#else
+#include <sqlite3.h>
+#endif
+#include <stdlib.h>
+#include <stdint.h>
+*/
+import "C"
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+	"unsafe"
+)
+
+const (
+	SQLITEDeserializeFreeOnClose = 1
+)
+
+// Serialize returns a byte slice that is a serialization of the database.
+//
+// See https://www.sqlite.org/c3ref/serialize.html
+func (c *SQLiteConn) Serialize(schema string) ([]byte, error) {
```

Comment:
@otoolep What has been your typical use case for this method? I am wondering if it would be useful to have a `SerializeTo` version that takes an `io.Writer` instead of returning `[]byte`. That way if your intent is to stream it somewhere (like a file or HTTP response) we don't have to have two copies in RAM simultaneously.

It could be nice to have something similar for `Deserialize` with `io.Reader`, but it would be more complicated because of the buffering that is required.

Reply:
Good idea, I would very much like to reduce memory usage. I've never been happy with the amount of memory copying going on in callers.

rqlite uses this method to snapshot the SQLite database to disk, and part of a process know as "Log Truncation". 

Serialization is triggered here: https://github.com/rqlite/rqlite/blob/697782280fc22ccf66c23693bb0a235538664257/store/store.go#L1347

The byte slice is then written to disk here: https://github.com/rqlite/rqlite/blob/697782280fc22ccf66c23693bb0a235538664257/store/store.go#L1394

The only detail is that the byte slice is first compressed before writing it to disk. I still think we could probably avoid the double copies.

Reply:
How about we get this PR in, and I follow up with a `SerializeTo()` call?

Reply:
That said, rqlite needs a copy of the SQLite database at a very specific instant, and that copy is not written immediately to disk (because the rqlite doesn't know the file path at the time the Snapshotting process asks for a copy of the SQLite database). So getting a copy of the database, and writing it to disk, happen at different times, possibly from different Goroutines.

So rqlite may not make immediately use of any `SerializeTo` but I'm going to think about it. I might be able to make it work.

Reply:
> How about we get this PR in, and I follow up with a SerializeTo() call?

Yes that sounds reasonable.